adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java

```diff
@@ -0,0 +1,72 @@
+package com.squareup.moshi.adapters;
+
+import com.squareup.moshi.Json;
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonDataException;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import java.io.IOException;
+import java.util.Arrays;
+import javax.annotation.Nullable;
+
+/**
+ * A JsonAdapter for enums that allows having a fallback enum value when a deserialized string does
+ * not match any enum value.
+ */
+public final class ConfigurableEnumJsonAdapter<T extends Enum<T>> extends JsonAdapter<T> {
+  final Class<T> enumType;
+  final String[] nameStrings;
+  final T[] constants;
+  final JsonReader.Options options;
+  final @Nullable T fallbackValue;
+
+  public static <T extends Enum<T>> ConfigurableEnumJsonAdapter<T> create(Class<T> enumType) {
+    return new ConfigurableEnumJsonAdapter<>(enumType, null);
+  }
+
+  public ConfigurableEnumJsonAdapter<T> withFallbackValue(T fallbackValue) {
```

Comment:
Should document when we fallback? Like only when an enum is unknown. (But not absent or null!)

Maybe should be `withUnknownValue()` ?

Reply:
Might be extra worth a warning to say that this adapter is case-sensitive and that the fallback will be used even on just a case mismatch

Reply:
oops, i missed putting in a doc. will update.

`withUnknownValue` reads strange to me. it sounds like the value passed in is unknown, when it is really the substitute for unknown values.

Reply:
Maybe withUnknownFallback() ?

Reply:
sounds good.